In [1]:
import networkx as nx
import matplotlib.pyplot as plt

# On importe un graphe pour chacune des université
G_caltech = nx.read_gml('./db/Caltech36.gml')
G_mit = nx.read_gml('./db/MIT8.gml')
G_hopkins = nx.read_gml('./db/Johns Hopkins55.gml')

# Question 4
On cherche désormais à prédire des liens.

In [2]:
import numpy as np 
from link_prediction import generate_non_edges, JaccardCoefficient, CommonNeighbors, AdamicAdar

fractions = [0.05, 0.1, 0.15, 0.2]
k_values = [50, 100, 200, 400]
f, k = 0.1, 50

In [3]:
def remove_random_edges(g, fraction):
    mask = np.random.randint(0, 100, len(g.edges())) < fraction * 100
    e_removed = np.array(g.edges())[mask]
    g_train = g.copy()
    g_train.remove_edges_from(e_removed)
    return g_train, e_removed

In [4]:
def generate_precision(f, k):
    # On commence par supprimer aléatoirement des arêtes du graphe étudié
    ## g_train, e_removed = remove_random_edges(G_caltech, f)
    g_train, e_removed = remove_random_edges(G_mit, f)
    non_edges = generate_non_edges(g_train)

    # Prédiction de liens par l'indice de Jaccard
    jp = JaccardCoefficient(g_train)
    predictions_jp = jp.predict(non_edges)
    # Prédiction de liens par le nombre de voisins communs
    cn = CommonNeighbors(g_train)
    predictions_cn = cn.predict(non_edges)
    # Prédiction de liens par l'indice d'Adamic-Adar
    aa = AdamicAdar(g_train)
    predictions_aa = aa.predict(non_edges)
    precisions = []
    for p in [predictions_jp, predictions_cn, predictions_aa]: 
        predictions_sorted = np.argsort(p)[::-1][:k]
        e_removed_tuples = [(u, v) for u, v in zip(e_removed[:,0], e_removed[:,1])]
        
        tp = 0 # vrais positifs
        fp = 0 # faux positifs
        for x in predictions_sorted:
            if non_edges[x] in e_removed_tuples:
                tp += 1
            else:
                fp += 1
        
        precisions.append(tp / (tp + fp))
    return precisions

In [5]:
import itertools
for x, y in itertools.product(fractions, k_values):
    print(f"Précision pour f = {x}, k = {y} : {generate_precision(x, y)}")

Précision pour f = 0.05, k = 50 : [0.14, 0.62, 0.6]
Précision pour f = 0.05, k = 100 : [0.21, 0.54, 0.54]
Précision pour f = 0.05, k = 200 : [0.13, 0.485, 0.52]
Précision pour f = 0.05, k = 400 : [0.245, 0.4875, 0.5175]
Précision pour f = 0.1, k = 50 : [0.26, 0.72, 0.74]
Précision pour f = 0.1, k = 100 : [0.29, 0.77, 0.76]
Précision pour f = 0.1, k = 200 : [0.245, 0.765, 0.765]
Précision pour f = 0.1, k = 400 : [0.3325, 0.6675, 0.6925]
Précision pour f = 0.15, k = 50 : [0.24, 0.82, 0.82]
Précision pour f = 0.15, k = 100 : [0.34, 0.83, 0.83]
Précision pour f = 0.15, k = 200 : [0.32, 0.825, 0.835]
Précision pour f = 0.15, k = 400 : [0.4325, 0.7825, 0.795]
Précision pour f = 0.2, k = 50 : [0.38, 0.86, 0.88]
Précision pour f = 0.2, k = 100 : [0.26, 0.89, 0.89]
Précision pour f = 0.2, k = 200 : [0.35, 0.865, 0.865]
Précision pour f = 0.2, k = 400 : [0.4775, 0.8225, 0.8325]


## Résultats 
### pour Caltech
- Précision pour f = 0.05, k = 50 : [0.285, 0.48, 0.45]
- Précision pour f = 0.05, k = 100 : [0.31, 0.46, 0.44]
- Précision pour f = 0.05, k = 200 : [0.27, 0.375, 0.36]
- Précision pour f = 0.05, k = 400 : [0.21, 0.26, 0.24]
- Précision pour f = 0.1, k = 50 : [0.37, 0.68, 0.695]
- Précision pour f = 0.1, k = 100 : [0.355, 0.61, 0.6]
- Précision pour f = 0.1, k = 200 : [0.34, 0.53, 0.52]
- Précision pour f = 0.1, k = 400 : [0.35, 0.465, 0.45]
- Précision pour f = 0.15, k = 50 : [0.44, 0.76, 0.8]
- Précision pour f = 0.15, k = 100 : [0.51, 0.77, 0.75]
- Précision pour f = 0.15, k = 200 : [0.45, 0.66, 0.645]
- Précision pour f = 0.15, k = 400 : [0.48, 0.56, 0.54]
- Précision pour f = 0.2, k = 50 : [0.42, 0.88, 0.88]
- Précision pour f = 0.2, k = 100 : [0.535, 0.82, 0.83]
- Précision pour f = 0.2, k = 200 : [0.515, 0.71, 0.7]
- Précision pour f = 0.2, k = 400 : [0.52, 0.67, 0.66]
### pour MIT
- Précision pour f = 0.05, k = 50 : [0.14, 0.62, 0.6]
- Précision pour f = 0.05, k = 100 : [0.21, 0.54, 0.54]
- Précision pour f = 0.05, k = 200 : [0.13, 0.485, 0.52]
- Précision pour f = 0.05, k = 400 : [0.245, 0.4875, 0.5175]
- Précision pour f = 0.1, k = 50 : [0.26, 0.72, 0.74]
- Précision pour f = 0.1, k = 100 : [0.29, 0.77, 0.76]
- Précision pour f = 0.1, k = 200 : [0.245, 0.765, 0.765]
- Précision pour f = 0.1, k = 400 : [0.3325, 0.6675, 0.6925]
- Précision pour f = 0.15, k = 50 : [0.24, 0.82, 0.82]
- Précision pour f = 0.15, k = 100 : [0.34, 0.83, 0.83]
- Précision pour f = 0.15, k = 200 : [0.32, 0.825, 0.835]
- Précision pour f = 0.15, k = 400 : [0.4325, 0.7825, 0.795]
- Précision pour f = 0.2, k = 50 : [0.38, 0.86, 0.88]
- Précision pour f = 0.2, k = 100 : [0.26, 0.89, 0.89]
- Précision pour f = 0.2, k = 200 : [0.35, 0.865, 0.865]
- Précision pour f = 0.2, k = 400 : [0.4775, 0.8225, 0.8325]

### Interprétation
- Common Neighbors et Adamic-Adar sont bien plus efficaces que Jaccard pour prédire les liens manquants dans les graphes sociaux de Caltech et MIT. On remarque par ailleurs que les trois méthodes sont globalement plus précises quand le nombre d'arêtes supprimées f augmente et k diminue, à l'exception de la méthode de Jaccard qui fonctionne mieux à haut k sur le graphe de MIT.
  
- Par ailleurs, on retrouve que :
    - Caltech possède un réseau plus fragmenté (les performances sont plus faibles).
    - MIT possède un réseau plus dense, où les amis partagent souvent des cercles communs (précisions de Common Neighbors et Adamic-Adar plus élevées, surtout pour de faibles valeurs de k).